### DISCUSSION TITLE
Transform createWebStoragePersistor to createSyncStoragePersistor

### DISCUSSION THREAD

**Discussion by todorone** [TanStack/query#3667] [Ideas] [answered]
Currently RQ has `createWebStoragePersistor` and `createAsyncStoragePersistor` which assumes that non-web platforms can have async storages only, which is not true. React Native has performant synchronous MMKV adapters support, which can't be used with RQ right now directly(promisify adapter should be introduced).

The proposition is to rename `createWebStoragePersistor` => `createSyncStoragePersistor` and make it platform agnostic to cover RN sync providers. Looking at the code, this should be straightforward and easy change.

@TkDodo WDYT?

**Comment by TkDodo** [comment, marked as answer]
looks like we're currently using the `Storage` type from `lib.dom.d.ts`. Tihs is the type that `window.localStorage` and `window.sessionStorage` also have. Are the react native adapters not compatible with that type?

**Reply by todorone** [reply]
Nope, this type has strange `key()` method and `length` property.

And there is also the check for global `window`... :(

**Reply by TkDodo** [reply]
> And there is also the check for global window... :(

the global window check was removed in v4 (beta branch)

we can definitely just define our own types that minimally satisfy our required interface. I'm just not so sure about another last-minute rename 😅

**Reply by todorone** [reply]
Wow, great news, waiting for v4! 🎉 

Well, rename is not crucial, it just makes sense as i explained above.

**Reply by TkDodo** [reply]
okay, would you like to contribute the type definitions?
should be similar to the async ones, just ... not async I guess

https://github.com/tannerlinsley/react-query/blob/80958591fde797fd7040ee6fcf04c642e97d3008/src/createAsyncStoragePersister/index.ts#L5-L9

**Reply by todorone** [reply]
Yep, will do it on weekend if it's not too late...🐱

**Reply by TkDodo** [reply]
weekend is perfect 👍

**Reply by cloudorbush** [reply]
@TkDodo I'm getting the following error using the `createAsyncStoragePersister` in v4. 

```bash
Tried to access storage on the server. Did you forget to call this in useEffect?
```

Any way to implement the window check ourselves?

I'm using React Native Web and MMKV. 

Was working fine in v3.

Thanks!

**Reply by TkDodo** [reply]
How does your code look like?

**Reply by cloudorbush** [reply]
I just entered a bunch of ninja leet code and it seems to work now

here is my `persister.tsx` file

```jsx
import { MMKV } from 'react-native-mmkv'
import { createAsyncStoragePersister } from 'react-query/createAsyncStoragePersister'

const canUseDOM =
  typeof window !== 'undefined' && window.document?.createElement != null
const storage = new MMKV()

// create mmkv wrapper for asyncStorage
const valueOrNull = (value: any = null) => value

const mmkvStorageConfig = canUseDOM
  ? {
      setItem: (key: string, value: string): Promise<void> => {
        storage.set(key, value)
        return Promise.resolve()
      },
      getItem: (key: string): Promise<string | null> => {
        const value = valueOrNull(storage.getString(key))
        return Promise.resolve(value)
      },
      removeItem: (key: string): Promise<void> => {
        storage.delete(key)
        return Promise.resolve()
      },
    }
  : undefined

export const persister = createAsyncStoragePersister({
  storage: mmkvStorageConfig,
})

```

**Reply by todorone** [reply]
@vbylen In RQ v4 there is no need to create async storage anymore. More performant way is `createWebStoragePersister()`. The name is confusing though as it has nothing to do with web now(no `window` check + no web API), it's synchronous persister.

**Reply by TkDodo** [reply]
yeah I'm aware. let's see if we do a last minute rename 🤔

**Reply by TkDodo** [reply]
@todorone please create a PR that renames it to `SyncStoragePersister` :)

**Reply by todorone** [reply]
@TkDodo Sure, hopefully will do it today or tomorrow...

**Reply by TkDodo** [reply]
tomorrow is fine. Tanner is doing some restructurings right now, so might be best to wait until that's done. Please be aware to use the `beta` branch.

**Reply by todorone** [reply]
Seems it could be more sophisticated change than i thought - examples + docs + builds(?), not sure i'll have enough time to dive into all of this.
I created draft PR, please let me know if something particular should be adjusted, or if some drastic changes needed, maybe somebody else could pick it up.

https://github.com/TanStack/query/pull/3693/files
